Block Supports: Extract reusable preset input control from spacing sizes and border radius controls

aaronrobertshaw opened the pull request:
Part of:
- https://github.com/WordPress/gutenberg/issues/73772

Follow up to:
- https://github.com/WordPress/gutenberg/pull/71905

## What?

Extracts a resuable control component for a `PresetInputControl` that offers a segmented slider for choosing presets, a custom value toggle, and the ability to display a CustomSelectControl instead of the custom RangeControl slider if there are too many presets.


## Why?

More and more block supports are going to be using presets and share a similar UI. Currently, we already have Padding, Margin, and Border Radius block supports using this type of control.

Extracting this into a new reusable component will allow us to clean up the code for the existing block support UIs (`SpacingSizesControl` and `BorderRadiusControl`), and then leverage it across a range of new and existing block supports that need preset functionality. These include:

- Min height
- Min width
- Width
- Height

## How?

- Creates a new control called `PresetInputControl`
- `PresetInputControl` aims to satisfy the slightly differing needs of the existing spacing and border radius controls
- Refactored `SpacingSizesControl` to use new preset input control
- Refactored `BorderRadiusControl` to use new preset input control
- Added some tests for both existing controls before refactor

## Testing Instructions

Test the existing controls for padding, margin, and border radius. Make sure that:
- Presets can be selected
- Custom values can be applied via toggle
- When excessive presets are in theme.json the control displays custom select control 
- Values can be applied and saved correctly
- All global styles and block support controls still function as expected

aaronrobertshaw committed:
Fix docblock

aaronrobertshaw committed:
Draft PresetInputControl

aaronrobertshaw committed:
Add unit tests for PrestInputControl utils

aaronrobertshaw committed:
Fix old typo in CUSTOM_VALUE_SETTINGS

aaronrobertshaw committed:
Add tests for BorderRadiusControl prior to refactor

aaronrobertshaw committed:
Refactor BorderRadiusControl to use PresetInputControl

aaronrobertshaw committed:
Add tests to SpacingSizesControl before refactor

aaronrobertshaw committed:
Refactor SpacingSizesControl to use new PresetInputControl

aaronrobertshaw committed:
Clean up PresetInputControl styling

aaronrobertshaw committed:
Fix

ramonjd commented on packages/block-editor/src/components/preset-input-control/index.js:
```diff
@@ -0,0 +1,367 @@
+		! isValuePreset( value, presetType ) &&
+		showCustomValueControl !== true
+	) {
+		setShowCustomValueControl( true );
```
Is this set state okay here? Or should it be in a useEffect?

aaronrobertshaw replied:
Done in https://github.com/WordPress/gutenberg/pull/73787/commits/44e0b3a00a84c276b0b59869484c1b5b4bbfff98

ramonjd commented on packages/block-editor/src/components/preset-input-control/index.js:
```diff
@@ -0,0 +1,367 @@
+	units,
+} ) {
+	// Treat value as a preset value if the passed in value matches the value of one of the presets.
+	value = getPresetValueFromCustomValue( value, presets, presetType );
```
Is it okay to mutate the value here?  I mean, it works, just thinking of later debugging and stable props.

Thinking of something like

```js
const normalizedValue = useMemo( 
    () => getPresetValueFromCustomValue( value, spacingSizes ),
    [ value, spacingSizes ]
);
```

I could be way off!

aaronrobertshaw replied:
Done in https://github.com/WordPress/gutenberg/pull/73787/commits/97de20dda3b34c5d36a6c1d088793bef058e5948

ramonjd commented on packages/block-editor/src/components/preset-input-control/constants.js:
```diff
@@ -0,0 +1,32 @@
+	vw: { max: 100, steps: 1 },
+	vh: { max: 100, steps: 1 },
+	em: { max: 10, steps: 0.1 },
+	rem: { max: 10, steps: 0.1 },
```
I just compared this with the list in in spacing-input-control.js and I see `rm: { max: 10, steps: 0.1 },` - that looks like a typo. 🤷🏻

aaronrobertshaw replied:
Yeah, it looks like it was in the spacing-input-control.js for a while. Figured the fix might as well land in this refactor.

ramonjd reviewed (commented):
Great work getting this up so quickly.

Looking good so far. 


https://github.com/user-attachments/assets/ca21dea5-9c82-41c3-a873-a701b423346f

I think the dropdown (with lots of border radius presets) positions itself outside the window border, but it also does on trunk.

aaronrobertshaw committed:
Fix mutating prop in new PresetInputControl

aaronrobertshaw committed:
Fix side effect during slider render phase

aaronrobertshaw committed:
Add baseline tests for the PresetInputControl

aaronrobertshaw committed:
Refactor PresetInputControl to use CustomValueControl

aaronrobertshaw commented:
I have:
- addressed the current feedback, 
- rebased on trunk, 
- refactored the custom value controls into a subcomponent, 
- and added some additional tests for the `PresetInputControl`

> I think the dropdown (with lots of border radius presets) positions itself outside the window border, but it also does on trunk.

I think this should be addressed via a separate follow-up.

Assuming that follow-up, I think this PR is ready for additional testing and review.

ramonjd reviewed (approved):
Nice work.

✅ Presets can be selected
✅ Custom values can be applied via toggle
✅ When excessive presets are in theme.json the control displays custom select control
✅ Values can be applied and saved correctly
✅ All global styles and block support controls still function as expected

<img width="272" height="537" alt="Screenshot 2025-12-11 at 2 52 21 pm" src="https://github.com/user-attachments/assets/024473d1-1c35-45a3-b7b1-2f1ed5ed7d8b" />
<img width="260" height="339" alt="Screenshot 2025-12-11 at 2 54 53 pm" src="https://github.com/user-attachments/assets/76d82386-01d1-4ad9-a25b-1a259d8f8237" />

There was a failing E2E related to the spinning button. It might be related. I kicked it off again just in case. 

Otherwise, LGTM

andrewserong commented:
Just a quick drive-by comment: I think the e2e test will need to be updated because the string in this PR has changed from `Set custom size` to `Set custom value`:

https://github.com/WordPress/gutenberg/blob/d47370d15091da91e3ef1b3bf228393db5ee8ff0/test/e2e/specs/site-editor/styles.spec.js#L63

Hopefully nothing more significant than that! 🤞

(And great work on the consolidation @aaronrobertshaw!)

aaronrobertshaw committed:
Fix e2e
